generate-icons.js

```diff
@@ -0,0 +1,158 @@
+import sharp from 'sharp';
+import path from 'path';
+import fs from 'fs/promises';
+import pngToIco from 'png-to-ico';
+import { fileURLToPath } from 'url';
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = path.dirname(__filename);
+
+const SETTINGS = {
+  outputDirectory: 'images',
+  iconLightFilename: 'images/src/crest-negative.svg',
+  iconDarkFilename: 'images/src/crest-positive.svg',
+  lightColor: '#fdfcfc',
+  darkColor: '#0c0c0c',
+  androidChromeIconSizes: [36, 48, 72, 96, 144, 192],
+  appleTouchIconSizes: [57, 60, 72, 76, 114, 120, 144, 152, 180],
+  faviconSizes: [16, 32, 96],
+}
+
+const transparent = {r: 0, g: 0, b: 0, alpha: 0};
+
+try {
+  SETTINGS.androidChromeIconSizes.forEach(async (size) => {
+    await generateIcon({
+      size,
+      padding: Math.floor(size / 20),
+      inputFilename: SETTINGS.iconLightFilename,
+      outputFilename: `android-chrome-${size}x${size}.png`,
+      background: transparent,
+    })
+  });
+
+  SETTINGS.appleTouchIconSizes.forEach(async (size) => {
+    await generateIcon({
+      size,
+      padding: Math.floor(size / 10),
+      inputFilename: SETTINGS.iconDarkFilename,
+      outputFilename: `apple-touch-icon-${size}x${size}.png`,
+      background: {r: 255, g: 255, b: 255, alpha: 1},
+    });
+  });
+
+  SETTINGS.faviconSizes.forEach(async (size) => {
+    await generateIcon({
+      size,
+      padding: 0,
+      inputFilename: SETTINGS.iconDarkFilename,
+      outputFilename: `favicon-${size}x${size}.png`,
+      background: transparent,
+    });
+  });
+
+  SETTINGS.faviconSizes.forEach(async (size) => {
+    await generateIcon({
+      size,
+      padding: 0,
+      inputFilename: SETTINGS.iconLightFilename,
+      outputFilename: `favicon-dark-mode-${size}x${size}.png`,
+      background: transparent,
+    });
+  });
+
+  const faviconBuffer = await pngToIco(SETTINGS.faviconSizes.map((size) => path.join(__dirname, SETTINGS.outputDirectory, `favicon-${size}x${size}.png`)));
+  await fs.writeFile(path.join(__dirname, SETTINGS.outputDirectory, 'favicon.ico'), faviconBuffer);
+
+  const faviconDarkBuffer = await pngToIco(SETTINGS.faviconSizes.map((size) => path.join(__dirname, SETTINGS.outputDirectory, `favicon-dark-mode-${size}x${size}.png`)));
+  await fs.writeFile(path.join(__dirname, SETTINGS.outputDirectory, 'favicon-dark-mode.ico'), faviconDarkBuffer);
+
+// MS Tile icons - these are more faffy as regards padding and sizes
```

Comment:
I think the `sharp` options could make this easier, not having to specify handcrafted padding. The `extend` method can add padding and the `resize` method has options deliberately very similar to CSS's `object-fit` property, for the non-square sizes where you want the image to be contained within.

Not worth changing now it's done, especially given how useless these MS images might be these days, but something to remember about sharp